File: backend/app/models/habit.py
from datetime import date, datetime, timezone

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Integer, JSON, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.database.database import Base


class Habit(Base):
    __tablename__ = "habits"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )

    name: Mapped[str] = mapped_column(String(120), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str] = mapped_column(String(50), default="personal", index=True)
    color: Mapped[str] = mapped_column(String(20), default="#4a6ded")
    icon: Mapped[str] = mapped_column(String(40), default="sparkles")

    frequency: Mapped[str] = mapped_column(String(20), default="daily", index=True)
    scheduled_days: Mapped[list[int]] = mapped_column(JSON, default=list)
    target_count: Mapped[int] = mapped_column(Integer, default=1)
    unit: Mapped[str] = mapped_column(String(30), default="times")

    reminder_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    reminder_time: Mapped[str | None] = mapped_column(String(5), nullable=True)

    start_date: Mapped[date] = mapped_column(Date, default=date.today)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    is_archived: Mapped[bool] = mapped_column(Boolean, default=False, index=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )


class HabitCompletion(Base):
    __tablename__ = "habit_completions"
    __table_args__ = (
        UniqueConstraint("habit_id", "completion_date", name="uq_habit_completion_date"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    habit_id: Mapped[int] = mapped_column(
        ForeignKey("habits.id", ondelete="CASCADE"), index=True, nullable=False
    )
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    completion_date: Mapped[date] = mapped_column(Date, index=True, nullable=False)
    count: Mapped[int] = mapped_column(Integer, default=1)
    note: Mapped[str | None] = mapped_column(String(300), nullable=True)
    completed_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )

File: frontend/types/habit.ts
export type HabitFrequency = "daily" | "weekdays" | "weekly" | "custom";
export type HabitCategory =
  | "health"
  | "study"
  | "work"
  | "mindfulness"
  | "fitness"
  | "personal";

export type Habit = {
  id: number;
  user_id: number;
  name: string;
  description: string | null;
  category: HabitCategory;
  color: string;
  icon: string;
  frequency: HabitFrequency;
  scheduled_days: number[];
  target_count: number;
  unit: string;
  reminder_enabled: boolean;
  reminder_time: string | null;
  start_date: string;
  end_date: string | null;
  is_archived: boolean;
  current_streak: number;
  best_streak: number;
  completed_today: boolean;
  today_count: number;
  completion_rate: number;
  created_at: string;
  updated_at: string;
};

export type HabitCompletion = {
  id: number;
  habit_id: number;
  completion_date: string;
  count: number;
  note: string | null;
  completed_at: string;
};

export type HabitWorkspace = {
  habits: Habit[];
  completions: HabitCompletion[];
  today_completed: number;
  today_total: number;
  weekly_rate: number;
  longest_streak: number;
};

export type HabitPayload = Omit<
  Habit,
  | "id"
  | "user_id"
  | "current_streak"
  | "best_streak"
  | "completed_today"
  | "today_count"
  | "completion_rate"
  | "created_at"
  | "updated_at"
>;
